internal/vcr/.semgrep-vcr.yml

```diff
@@ -141,3 +141,65 @@ rules:
     paths:
       include:
         - "**/*_test.go"
+
+  # Use the internal/retry package in place of the Plugin SDK V2 helper library.
+  # The internal variants of the state refresh functions are VCR-compatible.
+  - id: use-internal-retry
+    message: "Use internal/retry for helpers which are compatible with VCR acceptance testing"
+    languages: [go]
+    severity: WARNING
+    pattern-either:
+      - pattern: |
+          import "github.com/hashicorp/terraform-plugin-sdk/v2/helper/retry"
+      - pattern: |
+          import $ALIAS "github.com/hashicorp/terraform-plugin-sdk/v2/helper/retry"
+    fix-regex:
+      regex: 'github\.com/hashicorp/terraform-plugin-sdk/v2/helper/retry'
+      replacement: 'github.com/hashicorp/terraform-provider-aws/internal/retry'
+
+  - id: internal-retry-staterefreshfunc-pass-context
+    languages: [go]
+    message: "The internal retry.StateRefreshFunc accepts a context argument"
+    severity: WARNING
+    patterns:
+      - pattern-inside: "func $F(...) retry.StateRefreshFunc{ ... }"
+      - pattern: |
+          return func() (any, string, error) {
+            $...BODY
+          }
+    fix: |
+      return func(ctx context.Context) (any, string, error) {
+        $...BODY
+      }
+    paths:
+      include:
+        - "**/*.go"
+
+  - id: internal-retry-notfounderror-drop-lastrequest
+    languages: [go]
+    message: "The internal retry.NotFoundError struct drops the LastRequest argument"
+    severity: WARNING
+    pattern: |
+        retry.NotFoundError {
+          $...ARGS,
+          LastRequest: ...,
+        }
+    fix: |
+      retry.NotFoundError {
+        $...ARGS,
+      }
+    paths:
+      include:
+        - "**/*.go"
+
+  - id: internal-retry-notfound
```

Comment:
Rather than having this rule I think we should update `tfresource.NotFound` to also look for `internal/retry/NotFound` (plus I think there's something similar for `SetLastError`?).

Reply:
I tried that first but it causes and import cycle because some of the wrappers in the `retry` package reference `tfresource.NotFound`. I'll think on this, perhaps we need to adjust the package structure(s) a bit.

Reply:
After chatting on this more, we've opted to keep the `retry.NotFound` helper and alias `tfresource.NotFound` to it. The `retry` package will remove references to `tfresource` in favor of the local package variants, avoiding the import cycle. Lastly, this semgrep rule can be dropped as both helper functions will be pointing to the same underlying logic and migration is not yet necessary.